pyomo/contrib/pynumero/interfaces/nlp_projections.py

```diff
@@ -0,0 +1,303 @@
+        """
+        This class takes an NLP that depends on a set of primals (original
+        space) and converts it to an NLP that depends on a reordered set of 
+        primals (projected space).
+
+        This will impact all the returned items associated with primal
+        variables. E.g., the gradient will be in the new primals ordering
+        instead of the original primals ordering.
+
+        Note also that this can include additional primal variables not
+        in the original NLP, or can exclude primal variables that were 
+        in the original NLP.
+
+        Parameters
+        ----------
+        original_nlp : NLP-like
+            The original NLP object that implements the NLP interface
+
+        primals_ordering: list
+           List of strings indicating the desired primal variable 
+           ordering for this NLP. The list can contain new variables
+           that are not in the original NLP, thereby expanding the 
+           space of the primal variables.
+        """
+        super(ProjectedNLP, self).__init__(original_nlp)
+        self._primals_ordering = list(primals_ordering)
+        self._original_idxs = None
+        self._projected_idxs = None
+        self._generate_maps()
+        self._projected_primals = self.init_primals()
+        self._jacobian_nz_mask = None
+        self._hessian_nz_mask = None
+        self._nnz_jacobian = None
+        self._nnz_hessian_lag = None
+
+    def _generate_maps(self):
+        if self._original_idxs is None or self._projected_idxs is None:
+            primals_ordering_dict = {k:i for i,k in enumerate(self._primals_ordering)}
+            original_names = self._original_nlp.primals_names()
+            original_idxs = list()
+            projected_idxs = list()
+            for i,nm in enumerate(original_names):
+                if nm in primals_ordering_dict:
+                    # we need the reordering for this element
+                    original_idxs.append(i)
+                    projected_idxs.append(primals_ordering_dict[nm])
+            self._original_idxs = np.asarray(original_idxs)
+            self._projected_idxs = np.asarray(projected_idxs)
+            self._original_to_projected = np.nan*np.zeros(self._original_nlp.n_primals())
+            self._original_to_projected[self._original_idxs] = self._projected_idxs
+
+    def n_primals(self):
+        return len(self._primals_ordering)
+
+    def primals_names(self):
+        return list(self._primals_ordering)
+
+    def nnz_jacobian(self):
+        if self._nnz_jacobian is None:
+            J = self.evaluate_jacobian()
+            self._nnz_jacobian = len(J.data)
+        return self._nnz_jacobian
+
+    def nnz_hessian_lag(self):
+        if self._nnz_hessian_lag is None:
+            H = self.evaluate_hessian_lag()
+            self._nnz_hessian_lag = len(H.data)
+        return self._nnz_hessian_lag
+
+    def _project_primals(self, default, original_primals):
```

Comment:
You might consider renaming this function, as you really use it to project columns (e.g., in `evaluate_grad_objective` it is projecting the objective gradient terms and not the primal variable (values).)

Reply:
In evaluate_grad_objective, this is the gradient with respect to the primals. Therefore this projection is consistent. It is projected anything in the primal variable space.